File: modules/day-03-reproducible/demos/project-r-renv/summarize_model_counts.R
library(dplyr)

gene_set_uploads <- data.frame(
  created = c("01/10/2021", "03/15/2021", "07/02/2022", "04/21/2022", "08/08/2023", "09/14/2023"),
  model = c("mouse", "cell line", "mouse", "human", "human", "mouse"),
  search_text = c("opioid", "opioid", "opioid", "opioid", "opioid", "opioid")
)

summarize_model_counts <- function(data, start_year = "2021", end_year = "2023") {
  prepared <- data
  prepared$created <- as.Date(prepared$created, format = "%m/%d/%Y")
  prepared$year <- format(prepared$created, "%Y")

  filtered <- prepared[prepared$year >= start_year & prepared$year <= end_year, ]

  filtered %>%
    count(year, model) %>%
    arrange(year, model)
}

write_results <- function(result_table, output_file = "results/model_counts.csv") {
  output_dir <- dirname(output_file)

  if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
  }

  write.csv(result_table, output_file, row.names = FALSE)
}

main <- function() {
  result_table <- summarize_model_counts(gene_set_uploads)
  output_file <- "results/model_counts.csv"
  write_results(result_table, output_file)
  cat("Saved results to", output_file, "\n")
}

if (sys.nframe() == 0) {
  main()
}

File: modules/day-03-reproducible/demos/project-python-uv/summarize_model_counts.py
from pathlib import Path

import pandas as pd

gene_set_uploads = pd.DataFrame(
    {
        "created": [
            "01/10/2021",
            "03/15/2021",
            "07/02/2022",
            "04/21/2022",
            "08/08/2023",
            "09/14/2023",
        ],
        "model": ["mouse", "cell line", "mouse", "human", "human", "mouse"],
        "search_text": ["opioid", "opioid", "opioid", "opioid", "opioid", "opioid"],
    }
)


def summarize_model_counts(data, start_year="2021", end_year="2023"):
    prepared = data.copy()
    prepared["created"] = pd.to_datetime(prepared["created"], format="%m/%d/%Y")
    prepared["year"] = prepared["created"].dt.strftime("%Y")
    filtered = prepared[
        (prepared["year"] >= start_year) & (prepared["year"] <= end_year)
    ]
    return (
        filtered.groupby(["year", "model"])
        .size()
        .reset_index(name="n")
        .sort_values(["year", "model"])
        .reset_index(drop=True)
    )


def write_results(result_table, output_file):
    output_path = Path(output_file)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    result_table.to_csv(output_path, index=False)


def main():
    result_table = summarize_model_counts(gene_set_uploads)
    output_file = "results/model_counts.csv"
    write_results(result_table, output_file)
    print(f"Saved results to {output_file}")


if __name__ == "__main__":
    main()
